File: VB/CardDAVServer.FileSystemStorage.AspNet/Discovery.vb
Imports System
Imports System.Collections.Generic
Imports System.Threading.Tasks
Imports ITHit.WebDAV.Server
Imports ITHit.WebDAV.Server.CardDav
Imports CardDAVServer.FileSystemStorage.AspNet.CardDav

''' <summary>
''' Assists in finding folders that contain calendars and address books.
''' </summary>
Public Class Discovery
    Implements IAddressbookDiscovery

    ''' <summary>
    ''' Instance of <see cref="DavContext"/> .
    ''' </summary>
    Private context As DavContext

    Public Sub New(context As DavContext)
        Me.context = context
    End Sub

    ''' <summary>
    ''' Returns list of folders that contain address books owned by this principal.
    ''' </summary>
    ''' <remarks>This enables address books discovery owned by current loged-in principal.</remarks>
    Public Async Function GetAddressbookHomeSetAsync() As Task(Of IEnumerable(Of IItemCollection)) Implements IAddressbookDiscovery.GetAddressbookHomeSetAsync
        Dim addressbooksUserFolder As String = String.Format("{0}{1}/", AddressbooksRootFolder.AddressbooksRootFolderPath, context.UserName)
        Return {Await DavFolder.GetFolderAsync(context, addressbooksUserFolder)}
    End Function

    ''' <summary>
    ''' Returns <b>true</b> if <b>addressbook-home-set</b> feature is enabled, <b>false</b> otherwise.
    ''' </summary>
    ''' <remarks>
    ''' <para>
    ''' In this method you can analyze User-Agent header and enable/disable <b>addressbook-home-set</b> feature for specific client. 
    ''' </para>
    ''' <para>
    ''' iOS and OS X does require <b>addressbook-home-set</b> feature to be always enabled. On the other hand it may consume extra 
    ''' resources especially with iOS CardDAV client. iOS starts immediate synchronization of all address books found on the server 
    ''' via home-set request. Typically you will always enable heome-set for iOS and OS X CardDAV clients, but may disable it for other clients.
    ''' </para>
    ''' </remarks>
    Public ReadOnly Property AddressbookHomeSetEnabled As Boolean Implements IAddressbookDiscovery.AddressbookHomeSetEnabled
        Get
            Return True
        End Get
    End Property
End Class

File: CS/CardDAVServer.FileSystemStorage.AspNetCore/Discovery.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

using ITHit.WebDAV.Server;
using ITHit.WebDAV.Server.CardDav;
using CardDAVServer.FileSystemStorage.AspNetCore.CardDav;


namespace CardDAVServer.FileSystemStorage.AspNetCore
{
    /// <summary>
    /// Assists in finding folders that contain calendars and address books.
    /// </summary>
    public class Discovery : IAddressbookDiscovery
    {
        /// <summary>
        /// Instance of <see cref="DavContext"/>.
        /// </summary>
        private DavContext context;

        public Discovery(DavContext context)
        {
            this.context = context;
        }

        /// <summary>
        /// Returns list of folders that contain address books owned by this principal.
        /// </summary>
        /// <remarks>This enables address books discovery owned by current loged-in principal.</remarks>
        public async Task<IEnumerable<IItemCollection>> GetAddressbookHomeSetAsync()
        {
            string addressbooksUserFolder = string.Format("{0}{1}/", AddressbooksRootFolder.AddressbooksRootFolderPath, context.UserName);
            return new[] { await DavFolder.GetFolderAsync(context, addressbooksUserFolder) };
        }

        /// <summary>
        /// Returns <b>true</b> if <b>addressbook-home-set</b> feature is enabled, <b>false</b> otherwise.
        /// </summary>
        /// <remarks>
        /// <para>
        /// In this method you can analyze User-Agent header and enable/disable <b>addressbook-home-set</b> feature for specific client. 
        /// </para>
        /// <para>
        /// iOS and OS X does require <b>addressbook-home-set</b> feature to be always enabled. On the other hand it may consume extra 
        /// resources especially with iOS CardDAV client. iOS starts immediate synchronization of all address books found on the server 
        /// via home-set request. Typically you will always enable heome-set for iOS and OS X CardDAV clients, but may disable it for other clients.
        /// </para>
        /// </remarks>
        public bool AddressbookHomeSetEnabled
        {
            get
            {
                return true;
            }
        }
    }
}